File: Week 3/Week 3 Solutions/D/D_cutinline_bruteforce.py

# The brute force approach is to handle each query with a list.
# We can traverse the whole list, and process the person in the query when we find them
# This is O(N * Q * |S|) time, which is fine for this problem
# (N = # of people, Q = # of queries, |S| = length of each string)
# If we had harsher constraints, we would need to use the other approach

n = int(input())
line = []
for i in range(n):
    line.append(input())

q = int(input())
for i in range(q):
    query = input().split()
    if(query[0] == 'cut'):
        x, y = query[1], query[2]
        next_line = []
        for person in line:
            if(person == y):
                next_line.append(x)
            next_line.append(person)

        line = next_line
    else: # query[0] == 'leave'
        x = query[1]
        line.remove(x) # O(N) time

for person in line:
    print(person)

File: Week 3/Week 3 Solutions/D/D_cutinline_bruteforce.cpp
#include <bits/stdc++.h>
using namespace std;


int main(){
    // The brute force approach here is just to handle each query with a list.
    // We can traverse the whole list, and process the person in the query when we find them
    // This is O(N * Q * |S|) which is fine for this problem
    // (N = # of people, Q = # of queries, |S| = length of each string)
    // If we had harsher constraints, we would need to use the other approach
    
    int n;
    cin >> n;
    vector<string> line(n);
    for(int i = 0; i < n; i++){
        cin >> line[i];
    }
    int q;
    cin >> q;
    while(q-- > 0){
        string command;
        cin >> command;
        if(command == "cut"){
            string x,y;
            cin >> x >> y;
            vector<string> next_line;
            for(string s : line){
                if(s == y){
                    next_line.push_back(x);
                    next_line.push_back(y);
                }else{
                    next_line.push_back(s);
                }
            }
            line = next_line;
        }else{  // command == "leave"
            string x;
            cin >> x;
            line.erase(find(line.begin(), line.end(), x)); // Remove the person from the line; O(N) operation
                                                          
        }
    }
    for(string s : line){
        cout << s << '\n';
    }
}
